File: src/file_tools/client/_output.py
"""Bounded byte buffers used by command-execution clients."""

from __future__ import annotations

from dataclasses import dataclass, field


DEFAULT_MAX_OUTPUT_BYTES = 1024 * 1024
MAX_CONFIGURABLE_OUTPUT_BYTES = 16 * 1024 * 1024
OUTPUT_READ_CHUNK_BYTES = 64 * 1024
OUTPUT_DRAIN_TIMEOUT_SECS = 2.0
TERMINATION_GRACE_SECS = 1.0


def validate_max_output_bytes(value: int) -> int:
    """Validate the per-stream retained-output budget."""
    if isinstance(value, bool) or not isinstance(value, int):
        raise ValueError("max_output_bytes must be an integer")
    if value <= 0:
        raise ValueError("max_output_bytes must be greater than zero")
    if value > MAX_CONFIGURABLE_OUTPUT_BYTES:
        raise ValueError(
            "max_output_bytes must not exceed "
            f"{MAX_CONFIGURABLE_OUTPUT_BYTES} bytes"
        )
    return value


@dataclass
class HeadTailBytes:
    """Retain a stable prefix and rolling suffix while counting all bytes."""

    max_bytes: int = DEFAULT_MAX_OUTPUT_BYTES
    _head: bytearray = field(default_factory=bytearray, init=False, repr=False)
    _tail: bytearray = field(default_factory=bytearray, init=False, repr=False)
    total_bytes: int = field(default=0, init=False)

    def __post_init__(self) -> None:
        self.max_bytes = validate_max_output_bytes(self.max_bytes)

    @property
    def head_budget(self) -> int:
        return self.max_bytes // 2

    @property
    def tail_budget(self) -> int:
        return self.max_bytes - self.head_budget

    @property
    def retained_bytes(self) -> int:
        return len(self._head) + len(self._tail)

    @property
    def omitted_bytes(self) -> int:
        return max(0, self.total_bytes - self.retained_bytes)

    @property
    def truncated(self) -> bool:
        return self.omitted_bytes > 0

    def append(self, chunk: bytes) -> None:
        if not chunk:
            return
        self.total_bytes += len(chunk)

        head_remaining = self.head_budget - len(self._head)
        head_take = min(max(head_remaining, 0), len(chunk))
        if head_take:
            self._head.extend(chunk[:head_take])

        remainder = chunk[head_take:]
        if not remainder:
            return
        self._tail.extend(remainder)
        overflow = len(self._tail) - self.tail_budget
        if overflow > 0:
            del self._tail[:overflow]

    def raw_bytes(self) -> bytes:
        return bytes(self._head) + bytes(self._tail)

    def display_bytes(self) -> bytes:
        if not self.truncated:
            return self.raw_bytes()
        marker = (
            f"\n... [truncated, {self.omitted_bytes} bytes omitted] ...\n".encode()
        )
        return bytes(self._head) + marker + bytes(self._tail)

    def text(self) -> str:
        return self.display_bytes().decode("utf-8", errors="replace")

File: src/core-rs/output.rs
use std::cmp;

#[derive(Debug)]
pub struct HeadTailBytes {
    max_bytes: usize,
    head: Vec<u8>,
    tail: Vec<u8>,
    total_bytes: usize,
}

impl HeadTailBytes {
    pub fn new(max_bytes: usize) -> Self {
        Self {
            max_bytes,
            head: Vec::new(),
            tail: Vec::new(),
            total_bytes: 0,
        }
    }

    pub fn append(&mut self, chunk: &[u8]) {
        if chunk.is_empty() {
            return;
        }
        self.total_bytes = self.total_bytes.saturating_add(chunk.len());

        let head_budget = self.max_bytes / 2;
        let head_remaining = head_budget.saturating_sub(self.head.len());
        let head_take = cmp::min(head_remaining, chunk.len());
        self.head.extend_from_slice(&chunk[..head_take]);

        let remainder = &chunk[head_take..];
        if remainder.is_empty() {
            return;
        }
        let tail_budget = self.max_bytes - head_budget;
        self.tail.extend_from_slice(remainder);
        if self.tail.len() > tail_budget {
            let overflow = self.tail.len() - tail_budget;
            self.tail.drain(..overflow);
        }
    }

    pub fn total_bytes(&self) -> usize {
        self.total_bytes
    }

    pub fn retained_bytes(&self) -> usize {
        self.head.len() + self.tail.len()
    }

    pub fn omitted_bytes(&self) -> usize {
        self.total_bytes.saturating_sub(self.retained_bytes())
    }

    pub fn raw_bytes(&self) -> Vec<u8> {
        let mut result = Vec::with_capacity(self.retained_bytes());
        result.extend_from_slice(&self.head);
        result.extend_from_slice(&self.tail);
        result
    }

    pub fn display_bytes(&self) -> Vec<u8> {
        let omitted = self.omitted_bytes();
        if omitted == 0 {
            return self.raw_bytes();
        }
        let marker = format!("\n... [truncated, {omitted} bytes omitted] ...\n");
        let mut result = Vec::with_capacity(self.retained_bytes().saturating_add(marker.len()));
        result.extend_from_slice(&self.head);
        result.extend_from_slice(marker.as_bytes());
        result.extend_from_slice(&self.tail);
        result
    }
}
